File: lib/tac-events.ts
/**
 * TAC Intelligence event bus (Vercel local) — pipes ERP events into logs + AI insights.
 */

import { safeOpenRouterCall } from "@/lib/openrouter";
import { OpenRouterGuardError } from "@/lib/openrouter-guard";

export type TacInsight = {
  severity: "info" | "warning" | "anomaly";
  summary: string;
  actions: string[];
  source: "rule_engine" | "openrouter";
};

export type TacIntelligenceLog = {
  id: string;
  event_type: string;
  pillar_id: string;
  payload: Record<string, unknown>;
  insight: TacInsight;
  status: "logged";
  created_at: string;
};

const HIGH_VALUE_INVOICE = Number(
  process.env.FAOS_HIGH_VALUE_INVOICE_THRESHOLD || "1000"
);
const MAJOR_INVENTORY_DELTA = Number(
  process.env.FAOS_MAJOR_INVENTORY_DELTA || "25"
);
const MAX_LOGS = 200;

const intelligenceLogs: TacIntelligenceLog[] = [];

function now() {
  return new Date().toISOString();
}

function uid() {
  return `tac_intel_${Date.now().toString(36)}_${Math.random().toString(36).slice(2, 8)}`;
}

function ruleInsight(
  eventType: string,
  payload: Record<string, unknown>
): TacInsight {
  if (eventType === "order_finalized") {
    const total = Number(payload.total || 0);
    return {
      severity: total >= HIGH_VALUE_INVOICE ? "anomaly" : "info",
      summary: `Order ${payload.order_number} finalized — stock −${payload.quantity}, invoice linked, total ${payload.currency || "USD"} ${total.toFixed(2)}.`,
      actions: [
        "Verify warehouse pick completed",
        "Confirm invoice delivery to client",
      ],
      source: "rule_engine",
    };
  }
  if (eventType === "order_cancelled") {
    return {
      severity: "warning",
      summary: `Order ${payload.order_number} cancelled — stock restored and linked invoice marked cancelled.`,
      actions: ["Review cancellation reason", "Notify sales owner"],
      source: "rule_engine",
    };
  }
  if (eventType === "inventory_major_update") {
    const qty = Number(payload.quantity || 0);
    const reorder = Number(payload.reorder_level || 0);
    return {
      severity: qty <= reorder ? "anomaly" : "warning",
      summary: `Major inventory move on ${payload.sku}: delta ${Number(payload.delta || 0) >= 0 ? "+" : ""}${payload.delta}, on-hand ${qty} (reorder ${reorder}).`,
      actions: [
        qty <= reorder ? "Check reorder / PO pipeline" : "Audit stock movement",
      ],
      source: "rule_engine",
    };
  }
  if (eventType === "high_value_invoice") {
    return {
      severity: "anomaly",
      summary: `High-value invoice ${payload.invoice_number} (${payload.currency || "USD"} ${Number(payload.amount || 0).toFixed(2)}) for ${payload.client_name}.`,
      actions: ["Finance approval review", "Confirm payment terms"],
      source: "rule_engine",
    };
  }
  return {
    severity: "info",
    summary: String(payload.message || `TAC event recorded: ${eventType}`),
    actions: Array.isArray(payload.actions)
      ? (payload.actions as string[]).slice(0, 3)
      : [],
    source: "rule_engine",
  };
}

async function aiInsight(
  eventType: string,
  payload: Record<string, unknown>
): Promise<TacInsight | null> {
  if (!process.env.OPENROUTER_API_KEY?.trim()) return null;
  try {
    const result = await safeOpenRouterCall(
      [
        {
          role: "system",
          content:
            "Return only compact JSON with keys severity (info|warning|anomaly), summary, actions (max 3 strings).",
        },
        {
          role: "user",
          content: `FAOS TAC ERP event.\nevent_type=${eventType}\npayload=${JSON.stringify(payload)}`,
        },
      ],
      {
        maxTokens: 160,
        temperature: 0.2,
        clientKey: "tac-intelligence",
        intent: "analysis",
      }
    );
    let text = result.reply.trim();
    if (text.startsWith("```")) {
      text = text.replace(/^```(?:json)?\s*/i, "").replace(/\s*```$/, "");
    }
    const parsed = JSON.parse(text) as {
      severity?: TacInsight["severity"];
      summary?: string;
      actions?: string[];
    };
    return {
      severity: parsed.severity || "info",
      summary: parsed.summary || "AI insight generated.",
      actions: (parsed.actions || []).slice(0, 3),
      source: "openrouter",
    };
  } catch (err) {
    if (err instanceof OpenRouterGuardError) return null;
    return null;
  }
}

export async function emitIntelligenceEventLocal(
  eventType: string,
  payload: Record<string, unknown>,
  pillarId = "capital"
): Promise<TacIntelligenceLog> {
  const insight =
    (await aiInsight(eventType, payload)) || ruleInsight(eventType, payload);
  const record: TacIntelligenceLog = {
    id: uid(),
    event_type: eventType,
    pillar_id: pillarId,
    payload,
    insight,
    status: "logged",
    created_at: now(),
  };
  intelligenceLogs.push(record);
  if (intelligenceLogs.length > MAX_LOGS) {
    intelligenceLogs.splice(0, intelligenceLogs.length - MAX_LOGS);
  }
  return record;
}

export function listIntelligenceLogsLocal(limit = 50): TacIntelligenceLog[] {
  return [...intelligenceLogs]
    .sort((a, b) => b.created_at.localeCompare(a.created_at))
    .slice(0, Math.max(1, Math.min(limit, 200)));
}

export function shouldEmitInventoryEvent(
  delta: number,
  quantity: number,
  reorderLevel: number
): boolean {
  return Math.abs(delta) >= MAJOR_INVENTORY_DELTA || quantity <= reorderLevel;
}

export function shouldEmitInvoiceEvent(amount: number): boolean {
  return amount >= HIGH_VALUE_INVOICE;
}

export async function emitOrderSyncEventsLocal(
  order: {
    id: string;
    order_number: string;
    client_name: string;
    product_name: string;
    quantity: number;
    total: number;
    currency: string;
    status: string;
    invoice_id?: string;
    inventory_id?: string;
  },
  effects: {
    stock_adjusted: boolean;
    invoice_mutated: boolean;
    inventory: { id: string; sku: string; name: string; quantity: number; reorder_level: number } | null;
    invoice: { id: string; invoice_number: string; amount: number; currency: string; client_name: string } | null;
    events: Array<{ type: string; delta?: number }>;
  },
  cancelled = false
): Promise<string[]> {
  const ids: string[] = [];
  if (!effects.stock_adjusted && !effects.invoice_mutated) return ids;

  const log = await emitIntelligenceEventLocal(
    cancelled ? "order_cancelled" : "order_finalized",
    {
      order_id: order.id,
      order_number: order.order_number,
      client_name: order.client_name,
      product_name: order.product_name,
      quantity: order.quantity,
      total: order.total,
      currency: order.currency,
      status: order.status,
      invoice_id: order.invoice_id,
      inventory_id: order.inventory_id,
    },
    cancelled ? "capital" : "create"
  );
  ids.push(log.id);

  if (effects.invoice && shouldEmitInvoiceEvent(effects.invoice.amount)) {
    const hv = await emitIntelligenceEventLocal(
      "high_value_invoice",
      {
        invoice_id: effects.invoice.id,
        invoice_number: effects.invoice.invoice_number,
        amount: effects.invoice.amount,
        currency: effects.invoice.currency,
        client_name: effects.invoice.client_name,
        order_id: order.id,
      },
      "capital"
    );
    ids.push(hv.id);
  }

  if (effects.inventory && effects.stock_adjusted) {
    const delta =
      effects.events.find((e) => e.type.startsWith("inventory"))?.delta || 0;
    if (
      shouldEmitInventoryEvent(
        delta,
        effects.inventory.quantity,
        effects.inventory.reorder_level
      )
    ) {
      const invLog = await emitIntelligenceEventLocal(
        "inventory_major_update",
        {
          inventory_id: effects.inventory.id,
          sku: effects.inventory.sku,
          name: effects.inventory.name,
          delta,
          quantity: effects.inventory.quantity,
          reorder_level: effects.inventory.reorder_level,
          order_id: order.id,
        },
        "create"
      );
      ids.push(invLog.id);
    }
  }

  return ids;
}

export function tacLocalHealth() {
  return {
    status: "operational" as const,
    intelligence_logs: intelligenceLogs.length,
    last_event: intelligenceLogs[intelligenceLogs.length - 1]?.created_at ?? null,
  };
}

File: backend/services/tac_events.py
"""TAC Intelligence event bus — pipes ERP / system events into TAC logs + AI insights."""

from __future__ import annotations

import json
import os
import urllib.error
import urllib.request
import uuid
from datetime import datetime, timezone
from pathlib import Path
from typing import Any, Dict, List, Optional

STATE_PATH = Path(__file__).resolve().parents[1] / "data" / "faos_tac_state.json"

HIGH_VALUE_INVOICE = float(os.getenv("FAOS_HIGH_VALUE_INVOICE_THRESHOLD", "1000"))
MAJOR_INVENTORY_DELTA = int(os.getenv("FAOS_MAJOR_INVENTORY_DELTA", "25"))
MAX_INTELLIGENCE_LOGS = 200


def _now() -> str:
    return datetime.now(timezone.utc).isoformat()


def _load_state() -> Dict[str, Any]:
    if not STATE_PATH.exists():
        return {
            "last_sync": None,
            "pillar_tasks": [],
            "tac_commands": [],
            "intelligence_logs": [],
        }
    with STATE_PATH.open("r", encoding="utf-8") as f:
        data = json.load(f)
    data.setdefault("intelligence_logs", [])
    data.setdefault("tac_commands", [])
    data.setdefault("pillar_tasks", [])
    return data


def _save_state(data: Dict[str, Any]) -> None:
    from services.erp_tx import atomic_save

    atomic_save(STATE_PATH, data)


def _rule_insight(event_type: str, payload: Dict[str, Any]) -> Dict[str, Any]:
    """Deterministic management insight when OpenRouter is unavailable."""
    if event_type == "order_finalized":
        total = float(payload.get("total") or 0)
        qty = int(payload.get("quantity") or 0)
        severity = "anomaly" if total >= HIGH_VALUE_INVOICE else "info"
        return {
            "severity": severity,
            "summary": (
                f"Order {payload.get('order_number')} finalized — "
                f"stock −{qty}, invoice linked, total {payload.get('currency', 'USD')} {total:.2f}."
            ),
            "actions": [
                "Verify warehouse pick completed",
                "Confirm invoice delivery to client",
            ],
            "source": "rule_engine",
        }
    if event_type == "order_cancelled":
        return {
            "severity": "warning",
            "summary": (
                f"Order {payload.get('order_number')} cancelled — "
                "stock restored and linked invoice marked cancelled."
            ),
            "actions": ["Review cancellation reason", "Notify sales owner"],
            "source": "rule_engine",
        }
    if event_type == "inventory_major_update":
        delta = int(payload.get("delta") or 0)
        qty = int(payload.get("quantity") or 0)
        reorder = int(payload.get("reorder_level") or 0)
        severity = "anomaly" if qty <= reorder else "warning"
        return {
            "severity": severity,
            "summary": (
                f"Major inventory move on {payload.get('sku')}: delta {delta:+d}, "
                f"on-hand {qty} (reorder {reorder})."
            ),
            "actions": [
                "Check reorder / PO pipeline" if qty <= reorder else "Audit stock movement",
            ],
            "source": "rule_engine",
        }
    if event_type == "high_value_invoice":
        return {
            "severity": "anomaly",
            "summary": (
                f"High-value invoice {payload.get('invoice_number')} "
                f"({payload.get('currency', 'USD')} {float(payload.get('amount') or 0):.2f}) "
                f"for {payload.get('client_name')}."
            ),
            "actions": ["Finance approval review", "Confirm payment terms"],
            "source": "rule_engine",
        }
    if event_type == "system_state_change":
        return {
            "severity": "info",
            "summary": str(payload.get("message") or "System state changed."),
            "actions": list(payload.get("actions") or []),
            "source": "rule_engine",
        }
    return {
        "severity": "info",
        "summary": f"TAC event recorded: {event_type}",
        "actions": [],
        "source": "rule_engine",
    }


def _ai_insight(event_type: str, payload: Dict[str, Any]) -> Optional[Dict[str, Any]]:
    """Optional OpenRouter parse — never blocks the ERP transaction on failure."""
    api_key = os.getenv("OPENROUTER_API_KEY", "").strip()
    if not api_key:
        return None

    prompt = (
        "You are FAOS TAC Intelligence. Given this ERP event JSON, reply with a single JSON object "
        'with keys severity (info|warning|anomaly), summary (one sentence), actions (string array, max 3). '
        "No markdown.\n\n"
        f"event_type={event_type}\npayload={json.dumps(payload)}"
    )
    body = json.dumps(
        {
            "model": os.getenv("FAOS_TAC_INSIGHT_MODEL", "openai/gpt-4o-mini"),
            "messages": [
                {
                    "role": "system",
                    "content": "Return only compact JSON for management insights.",
                },
                {"role": "user", "content": prompt},
            ],
            "max_tokens": 180,
            "temperature": 0.2,
        }
    ).encode("utf-8")

    req = urllib.request.Request(
        "https://openrouter.ai/api/v1/chat/completions",
        data=body,
        headers={
            "Authorization": f"Bearer {api_key}",
            "Content-Type": "application/json",
            "HTTP-Referer": os.getenv(
                "NEXT_PUBLIC_SITE_URL", "https://faos-v5-0-workstation.vercel.app"
            ),
            "X-Title": "FAOS TAC Intelligence",
        },
        method="POST",
    )
    try:
        with urllib.request.urlopen(req, timeout=8) as resp:
            data = json.loads(resp.read().decode("utf-8"))
        text = (
            data.get("choices", [{}])[0]
            .get("message", {})
            .get("content", "")
            .strip()
        )
        if text.startswith("```"):
            text = text.strip("`")
            if text.startswith("json"):
                text = text[4:].strip()
        parsed = json.loads(text)
        return {
            "severity": parsed.get("severity") or "info",
            "summary": parsed.get("summary") or "AI insight generated.",
            "actions": list(parsed.get("actions") or [])[:3],
            "source": "openrouter",
        }
    except (urllib.error.URLError, TimeoutError, json.JSONDecodeError, KeyError, TypeError):
        return None


def emit_intelligence_event(
    event_type: str,
    payload: Dict[str, Any],
    *,
    pillar_id: str = "capital",
) -> Dict[str, Any]:
    """Append a TAC intelligence log and attach AI or rule-based insight."""
    insight = _ai_insight(event_type, payload) or _rule_insight(event_type, payload)
    record = {
        "id": f"tac_intel_{uuid.uuid4().hex[:12]}",
        "event_type": event_type,
        "pillar_id": pillar_id,
        "payload": payload,
        "insight": insight,
        "status": "logged",
        "created_at": _now(),
    }

    state = _load_state()
    logs: List[Dict[str, Any]] = state.setdefault("intelligence_logs", [])
    logs.append(record)
    if len(logs) > MAX_INTELLIGENCE_LOGS:
        state["intelligence_logs"] = logs[-MAX_INTELLIGENCE_LOGS:]
    state["last_sync"] = _now()
    _save_state(state)
    return record


def should_emit_inventory_event(delta: int, quantity: int, reorder_level: int) -> bool:
    return abs(delta) >= MAJOR_INVENTORY_DELTA or quantity <= reorder_level


def should_emit_invoice_event(amount: float) -> bool:
    return amount >= HIGH_VALUE_INVOICE


def list_intelligence_logs(limit: int = 50) -> List[Dict[str, Any]]:
    state = _load_state()
    logs = sorted(
        state.get("intelligence_logs", []),
        key=lambda r: r.get("created_at", ""),
        reverse=True,
    )
    return logs[: max(1, min(limit, 200))]


def module_health_snapshot(erp_counts: Dict[str, int]) -> Dict[str, Any]:
    state = _load_state()
    return {
        "tac": {
            "status": "operational",
            "last_sync": state.get("last_sync"),
            "commands": len(state.get("tac_commands", [])),
            "intelligence_logs": len(state.get("intelligence_logs", [])),
            "pillar_tasks": len(state.get("pillar_tasks", [])),
        },
        "modules": {
            "invoicing": {"status": "operational", "records": erp_counts.get("invoices", 0)},
            "inventory": {"status": "operational", "records": erp_counts.get("inventory", 0)},
            "hr": {"status": "operational", "records": erp_counts.get("employees", 0)},
            "orders": {"status": "operational", "records": erp_counts.get("orders", 0)},
            "products": {"status": "operational", "records": erp_counts.get("products", 0)},
        },
    }
